Bottom: 77e6f9a64fe21d866c09b5e3cf02a3b39e104bc2
Top:    2f5b9b6ed6a8de234fcae1bedfe828a856c38e4a
Author: Ian Kent <ikent@redhat.com>
Date:   2013-04-08 09:57:09 +0800

Due to the change in the expire-specific-submount-only patch, sub-mounts
within an indirect mount that follow a submount (in the check order) won't
be expired if that submount is busy.


---

diff --git a/lib/master.c b/lib/master.c
index a0e62f2..d769342 100644
--- a/lib/master.c
+++ b/lib/master.c
@@ -905,15 +905,32 @@ int master_notify_submount(struct autofs_point *ap, const char *path, enum state
 		this = list_entry(p, struct autofs_point, mounts);
 		p = p->prev;
 
-		if (!master_submount_list_empty(this)) {
-			mounts_mutex_unlock(ap);
-			return master_notify_submount(this, path, state);
-		}
-
 		/* path not the same */
 		if (strcmp(this->path, path))
 			continue;
 
+		if (!master_submount_list_empty(this)) {
+			/* char *this_path = strdup(this->path);
+			if (this_path) {
+				mounts_mutex_unlock(ap);
+				if (!master_notify_submount(this, path, state)) {
+					free(this_path);
+					ret = 0;
+					break;
+				}
+				mounts_mutex_lock(ap);
+				if (!__master_find_submount(ap, this_path)) {
+					free(this_path);
+					continue;
+				}
+				free(this_path);
+			} */
+			mounts_mutex_unlock(ap);
+			master_notify_submount(this, path, state);
+			mounts_mutex_lock(ap);
+			continue;
+		}
+
 		/* Now we have found the submount we want to expire */
 
 		st_mutex_lock();
@@ -959,10 +976,7 @@ int master_notify_submount(struct autofs_point *ap, const char *path, enum state
 			st_mutex_lock();
 		}
 		st_mutex_unlock();
-		mounts_mutex_unlock(ap);
-
-		return ret;
-
+		break;
 	}
 
 	mounts_mutex_unlock(ap);
